{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import anndata\n",
    "from scipy.stats import zscore\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rcParams[\"font.family\"] = \"Arial\"\n",
    "import matplotlib\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "import scanpy as sc\n",
    "\n",
    "import re\n",
    "\n",
    "from pyInfinityFlow.InfinityFlow_Utilities import marker_finder\n",
    "from pyInfinityFlow.InfinityFlow_Utilities import pearson_corr_df_to_df\n",
    "\n",
    "\n",
    "os.chdir(\"/media/kyle_storage/kyle_ferchen/grimes_lab_main/analysis/\"\\\n",
    "    \"2023_06_12_tea_seq_atac_processing/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Group</th>\n",
       "      <th>Old_order</th>\n",
       "      <th>CITE-to-TEA</th>\n",
       "      <th>Order</th>\n",
       "      <th>Level 1</th>\n",
       "      <th>Level 2</th>\n",
       "      <th>Level 3</th>\n",
       "      <th>Level Kairavee</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LT-HSC_Mllt3</td>\n",
       "      <td>HSCP</td>\n",
       "      <td>1</td>\n",
       "      <td>LT-HSC_Mllt3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>HSPC</td>\n",
       "      <td>HSC</td>\n",
       "      <td>qHSC</td>\n",
       "      <td>HSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ST-HSC</td>\n",
       "      <td>HSCP</td>\n",
       "      <td>2</td>\n",
       "      <td>ST-HSC</td>\n",
       "      <td>2.0</td>\n",
       "      <td>HSPC</td>\n",
       "      <td>HSC</td>\n",
       "      <td>aHSC</td>\n",
       "      <td>HSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MPP4-Hlf</td>\n",
       "      <td>HSCP</td>\n",
       "      <td>3</td>\n",
       "      <td>MPP4-Hlf</td>\n",
       "      <td>3.0</td>\n",
       "      <td>HSPC</td>\n",
       "      <td>MPP4</td>\n",
       "      <td>HSC-Ly</td>\n",
       "      <td>HSPC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MPP5-Egr1</td>\n",
       "      <td>HSCP</td>\n",
       "      <td>8</td>\n",
       "      <td>MPP5-Egr1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>HSPC</td>\n",
       "      <td>MPP5</td>\n",
       "      <td>MPP5-IER</td>\n",
       "      <td>MPP5-IER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MPP5-Flt3</td>\n",
       "      <td>HSCP</td>\n",
       "      <td>7</td>\n",
       "      <td>MPP5-Flt3</td>\n",
       "      <td>5.0</td>\n",
       "      <td>HSPC</td>\n",
       "      <td>MPP5</td>\n",
       "      <td>MPP5 Ly-I</td>\n",
       "      <td>HSPC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>ILC2</td>\n",
       "      <td>ILC</td>\n",
       "      <td>84</td>\n",
       "      <td>ILC2</td>\n",
       "      <td>84.0</td>\n",
       "      <td>ILC</td>\n",
       "      <td>ILC</td>\n",
       "      <td>ILC2</td>\n",
       "      <td>ILC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>Bcl11b+_preETP_Cd3d</td>\n",
       "      <td>ILC</td>\n",
       "      <td>85</td>\n",
       "      <td>Bcl11b+_preETP_Cd3d</td>\n",
       "      <td>85.0</td>\n",
       "      <td>T cell</td>\n",
       "      <td>preETP</td>\n",
       "      <td>pre-ILC1-ILC3-NKP</td>\n",
       "      <td>preETP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>Bcl11b+_preETP_Tdrd5</td>\n",
       "      <td>ILC</td>\n",
       "      <td>88</td>\n",
       "      <td>Bcl11b+_preETP_Tdrd5</td>\n",
       "      <td>86.0</td>\n",
       "      <td>T cell</td>\n",
       "      <td>preETP</td>\n",
       "      <td>pre-NKP</td>\n",
       "      <td>preETP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>ILC1-ILC3-NKP</td>\n",
       "      <td>ILC</td>\n",
       "      <td>87</td>\n",
       "      <td>ILC1-ILC3-NKP</td>\n",
       "      <td>87.0</td>\n",
       "      <td>ILC</td>\n",
       "      <td>ILC</td>\n",
       "      <td>ILC1-ILC3-NKP</td>\n",
       "      <td>ILC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>MEP-UNK</td>\n",
       "      <td>MultiLin</td>\n",
       "      <td>22</td>\n",
       "      <td>MEP-UNK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MultiLin</td>\n",
       "      <td>MEP</td>\n",
       "      <td>MEP-UNK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>88 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Cluster     Group  Old_order           CITE-to-TEA  Order  \\\n",
       "0           LT-HSC_Mllt3      HSCP          1          LT-HSC_Mllt3    1.0   \n",
       "1                 ST-HSC      HSCP          2                ST-HSC    2.0   \n",
       "2               MPP4-Hlf      HSCP          3              MPP4-Hlf    3.0   \n",
       "3              MPP5-Egr1      HSCP          8             MPP5-Egr1    4.0   \n",
       "4              MPP5-Flt3      HSCP          7             MPP5-Flt3    5.0   \n",
       "..                   ...       ...        ...                   ...    ...   \n",
       "83                  ILC2       ILC         84                  ILC2   84.0   \n",
       "84   Bcl11b+_preETP_Cd3d       ILC         85   Bcl11b+_preETP_Cd3d   85.0   \n",
       "85  Bcl11b+_preETP_Tdrd5       ILC         88  Bcl11b+_preETP_Tdrd5   86.0   \n",
       "86         ILC1-ILC3-NKP       ILC         87         ILC1-ILC3-NKP   87.0   \n",
       "87               MEP-UNK  MultiLin         22               MEP-UNK    NaN   \n",
       "\n",
       "     Level 1 Level 2            Level 3 Level Kairavee  \n",
       "0       HSPC     HSC               qHSC            HSC  \n",
       "1       HSPC     HSC               aHSC            HSC  \n",
       "2       HSPC    MPP4             HSC-Ly           HSPC  \n",
       "3       HSPC    MPP5           MPP5-IER       MPP5-IER  \n",
       "4       HSPC    MPP5          MPP5 Ly-I           HSPC  \n",
       "..       ...     ...                ...            ...  \n",
       "83       ILC     ILC               ILC2            ILC  \n",
       "84    T cell  preETP  pre-ILC1-ILC3-NKP         preETP  \n",
       "85    T cell  preETP            pre-NKP         preETP  \n",
       "86       ILC     ILC      ILC1-ILC3-NKP            ILC  \n",
       "87  MultiLin     MEP            MEP-UNK            NaN  \n",
       "\n",
       "[88 rows x 9 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_cluster_order = pd.read_csv(\"/media/kyle_storage/kyle_ferchen/\"\\\n",
    "    \"grimes_lab_main/data/2021_11_mouse_optimized_cite_seq/processed_files/\"\\\n",
    "    \"r7_cluster_order_kf_2024_01.csv\")\n",
    "map_r7_to_lv3 = pd.Series(\\\n",
    "    new_cluster_order[\"Level 3\"].values,\n",
    "    index=new_cluster_order[\"Cluster\"].values)\n",
    "map_lvl3_to_order = pd.Series(\\\n",
    "    new_cluster_order[\"Order\"].values,\n",
    "    index=new_cluster_order[\"Level 3\"].values)\n",
    "map_r7_to_replace_dash = pd.Series(\\\n",
    "    new_cluster_order[\"Cluster\"].values,\n",
    "    index=[i.replace(\"-\", \"_\") for i in new_cluster_order[\"Cluster\"].values])\n",
    "new_cluster_order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 55285 × 28692\n",
       "    obs: 'barcode', 'port', 'R7', 'lvl3', 'cluster_order'\n",
       "    var: 'gene', 'feature_type'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load CITE-seq data (SoupX 0.15 corrected)\n",
    "path_to_cite_data = \"/media/kyle_storage/kyle_ferchen/grimes_lab_main/data/\"\\\n",
    "    \"2021_11_mouse_optimized_cite_seq/processed_files/\"\n",
    "\n",
    "cite_adata = anndata.read_h5ad(os.path.join(\\\n",
    "    path_to_cite_data,\n",
    "    \"cite_seq_adata_rna_combined_SoupX_0_15_with_R7_clusters.h5ad\"))\n",
    "\n",
    "cite_adata.X = np.log2((10000 * (cite_adata.X.T / \\\n",
    "    cite_adata.X.sum(axis=1).T).T) + 1)\n",
    "\n",
    "cite_adata.obs[\"lvl3\"] = cite_adata.obs[\"R7\"].replace(\\\n",
    "    map_r7_to_lv3.to_dict()).values\n",
    "cite_adata.obs[\"cluster_order\"] = cite_adata.obs[\"lvl3\"].replace(\\\n",
    "    map_lvl3_to_order.to_dict()).values\n",
    "\n",
    "cite_adata = cite_adata[\\\n",
    "    cite_adata.obs.loc[cite_adata.obs[\"lvl3\"] != \"MEP-UNK\"].sort_values(\\\n",
    "        by=\"cluster_order\").index.values,\n",
    "    :]\n",
    "\n",
    "cite_df = pd.DataFrame(\\\n",
    "    cite_adata.X,\n",
    "    index=cite_adata.obs.index.values,\n",
    "    columns=cite_adata.var.index.values)\n",
    "\n",
    "cite_centroids = cite_df.copy()\n",
    "cite_centroids[\"cluster\"] = cite_adata.obs.loc[\\\n",
    "    cite_centroids.index.values, \"lvl3\"].values\n",
    "cite_centroids = pd.pivot_table(cite_centroids, index=\"cluster\", aggfunc=np.mean)\n",
    "\n",
    "cite_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Seqlet annotation\n",
    "seqlets = pd.read_feather(\"output/chrombpnet/modisco_merged_results/fold_0/\"\\\n",
    "    \"redo_extract_seqlets/all_seqlit_hits_above_modisco_min_anno.fea\")\n",
    "\n",
    "# Seqlets\n",
    "dp_scores = pd.read_feather(\"output/chrombpnet/modisco_merged_results/fold_0/\"\\\n",
    "    \"redo_extract_seqlets/all_seqlit_hits_above_modisco_min_dp_scores.fea\")\n",
    "dp_scores = dp_scores.set_index(\"index\").astype(np.float32)\n",
    "dp_scores = dp_scores.rename(map_r7_to_replace_dash.to_dict(), axis=1)\n",
    "dp_scores = dp_scores.rename(map_r7_to_lv3.to_dict(), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>5881540</td>\n",
       "      <td>7341540</td>\n",
       "      <td>chr10:3005001-4465000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>5321540</td>\n",
       "      <td>5881540</td>\n",
       "      <td>chr10:4465001-5025000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>4401540</td>\n",
       "      <td>5321540</td>\n",
       "      <td>chr10:5025001-5945000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>3961540</td>\n",
       "      <td>4401540</td>\n",
       "      <td>chr10:5945001-6385000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>3100000</td>\n",
       "      <td>3961540</td>\n",
       "      <td>chr10:6385001-7265000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2511</th>\n",
       "      <td>chrX</td>\n",
       "      <td>162907068</td>\n",
       "      <td>164497068</td>\n",
       "      <td>chrX:159345001-160935000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2512</th>\n",
       "      <td>chrX</td>\n",
       "      <td>164497068</td>\n",
       "      <td>166137068</td>\n",
       "      <td>chrX:160935001-162575000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2513</th>\n",
       "      <td>chrX</td>\n",
       "      <td>166137068</td>\n",
       "      <td>167267068</td>\n",
       "      <td>chrX:162575001-163705000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2514</th>\n",
       "      <td>chrX</td>\n",
       "      <td>167267068</td>\n",
       "      <td>168537068</td>\n",
       "      <td>chrX:163705001-164975000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2515</th>\n",
       "      <td>chrX</td>\n",
       "      <td>168537068</td>\n",
       "      <td>169337068</td>\n",
       "      <td>chrX:164975001-165775000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2516 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr      start        end                      name  score\n",
       "0     chr10    5881540    7341540     chr10:3005001-4465000      1\n",
       "1     chr10    5321540    5881540     chr10:4465001-5025000      1\n",
       "2     chr10    4401540    5321540     chr10:5025001-5945000      1\n",
       "3     chr10    3961540    4401540     chr10:5945001-6385000      1\n",
       "4     chr10    3100000    3961540     chr10:6385001-7265000      1\n",
       "...     ...        ...        ...                       ...    ...\n",
       "2511   chrX  162907068  164497068  chrX:159345001-160935000      1\n",
       "2512   chrX  164497068  166137068  chrX:160935001-162575000      1\n",
       "2513   chrX  166137068  167267068  chrX:162575001-163705000      1\n",
       "2514   chrX  167267068  168537068  chrX:163705001-164975000      1\n",
       "2515   chrX  168537068  169337068  chrX:164975001-165775000      1\n",
       "\n",
       "[2516 rows x 5 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in the TADs\n",
    "tads = pd.read_table(\"input/GSE119347_BMHSC_TADs/GSE119347_BMHSC_TADs_mm10_liftover.bed\", \n",
    "    header=None)\n",
    "tads.columns = [\"chr\", \"start\", \"end\", \"name\", \"score\"]\n",
    "tads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in the ATAC data\n",
    "path_atac_data = \"output/tea_r7_pseudobulk_from_pmat/\"\n",
    "\n",
    "peaks = pd.read_table(\\\n",
    "    os.path.join(path_atac_data, \"r7_tea_cluster_peak_set_with_tss.bed\"),\n",
    "    header=None)\n",
    "peaks.columns = [\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\"]\n",
    "\n",
    "atac_counts = pd.read_csv(\\\n",
    "    os.path.join(path_atac_data, \"r7_tea_pbulk_cpm_from_binary_pmat.csv\"))\n",
    "\n",
    "path_split_bams = \"output/tea_r7_split_bams/\"\n",
    "\n",
    "# Get the tea-seq metrics\n",
    "cell_counts = {}\n",
    "read_counts = {}\n",
    "for item in os.listdir(path_split_bams):\n",
    "    tmp_metrics = pd.read_table(os.path.join(path_split_bams, item,\n",
    "         \"metrics_bam_cluster_splits.txt\"))\n",
    "    for i, row in tmp_metrics.iterrows():\n",
    "        if row[\"cluster\"] in cell_counts:\n",
    "            cell_counts[row[\"cluster\"]] += row[\"cells_per_cluster\"]\n",
    "            read_counts[row[\"cluster\"]] += row[\"reads_per_cluster\"]\n",
    "        else:\n",
    "            cell_counts[row[\"cluster\"]] = row[\"cells_per_cluster\"]\n",
    "            read_counts[row[\"cluster\"]] = row[\"reads_per_cluster\"]\n",
    "\n",
    "tea_metrics = pd.concat([\\\n",
    "    pd.Series(cell_counts), \n",
    "    pd.Series(read_counts)], axis=1)\n",
    "tea_metrics.columns = [\"cells\", \"reads\"]\n",
    "tea_metrics.head()\n",
    "\n",
    "# Only select clusters with at least 100 cells\n",
    "sel_clusters = tea_metrics.loc[tea_metrics[\"cells\"] >= 100].index.values\n",
    "atac_counts = atac_counts[sel_clusters]\n",
    "\n",
    "# Get max log2 cpm observed for each peak\n",
    "peaks_to_index = pd.DataFrame({\\\n",
    "    \"i\": peaks.index.values,\n",
    "    \"peak\": (peaks[\"chr\"] + \":\" + peaks[\"start\"].astype(str) + \"-\" + \\\n",
    "        peaks[\"end\"].astype(str)).values,\n",
    "    \"max_cpm\": atac_counts.max(axis=1),\n",
    "    \"top_cluster\": atac_counts.idxmax(axis=1)})\n",
    "\n",
    "# Only accept peaks with a log2-cpm value > 3\n",
    "expressed_peaks = peaks_to_index.loc[\\\n",
    "    peaks_to_index[\"max_cpm\"] > 3][\"peak\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>Xkr4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3671497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373</th>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>chr1</td>\n",
       "      <td>4785709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>382</th>\n",
       "      <td>Lypla1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>4807822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>404</th>\n",
       "      <td>Tcea1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>4857813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>442</th>\n",
       "      <td>Rgs20</td>\n",
       "      <td>chr1</td>\n",
       "      <td>5018734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>751317</th>\n",
       "      <td>Uty</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1245690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>751334</th>\n",
       "      <td>Ddx3y</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1286628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>751345</th>\n",
       "      <td>Usp9y</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1459781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>757491</th>\n",
       "      <td>Gm21860</td>\n",
       "      <td>chrY</td>\n",
       "      <td>90755466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>757503</th>\n",
       "      <td>Gm47283</td>\n",
       "      <td>chrY</td>\n",
       "      <td>90784737</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18261 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene   chr     start\n",
       "114        Xkr4  chr1   3671497\n",
       "373      Mrpl15  chr1   4785709\n",
       "382      Lypla1  chr1   4807822\n",
       "404       Tcea1  chr1   4857813\n",
       "442       Rgs20  chr1   5018734\n",
       "...         ...   ...       ...\n",
       "751317      Uty  chrY   1245690\n",
       "751334    Ddx3y  chrY   1286628\n",
       "751345    Usp9y  chrY   1459781\n",
       "757491  Gm21860  chrY  90755466\n",
       "757503  Gm47283  chrY  90784737\n",
       "\n",
       "[18261 rows x 3 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Pull out the TSS positions from peaks\n",
    "tss_loci = peaks.loc[peaks[\"name\"].str.contains(\"TSS:\")].copy()\n",
    "tss_loci[\"gene\"] = [i.split(\":\")[-1] for i in tss_loci[\"name\"].values]\n",
    "tss_loci = tss_loci[[\"gene\", \"chr\", \"start\"]]\n",
    "tss_loci[\"start\"] = tss_loci[\"start\"] + 500\n",
    "tss_loci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Assigning TADs to ATAC peaks...\n",
      "\tWorking on chr10...\n",
      "\tWorking on chr11...\n",
      "\tWorking on chr12...\n",
      "\tWorking on chr19...\n",
      "\tWorking on chrX...\n",
      "\tWorking on chr8...\n",
      "\tWorking on chr13...\n",
      "\tWorking on chr5...\n",
      "\tWorking on chr2...\n",
      "\tWorking on chr6...\n",
      "\tWorking on chr14...\n",
      "\tWorking on chr15...\n",
      "\tWorking on chr18...\n",
      "\tWorking on chr16...\n",
      "\tWorking on chr17...\n",
      "\tWorking on chr4...\n",
      "\tWorking on chr1...\n",
      "\tWorking on chr3...\n",
      "\tWorking on chr7...\n",
      "\tWorking on chr9...\n",
      "\tWorking on chrX_GL456233_random...\n",
      "\tWorking on chrY...\n"
     ]
    }
   ],
   "source": [
    "## Assign each atac peak a TAD\n",
    "print(\"Assigning TADs to ATAC peaks...\")\n",
    "atac_tad_assignments = pd.Series([np.nan] * peaks.shape[0], \n",
    "    index=peaks.index.values)\n",
    "for tmp_chr in tads[\"chr\"].unique():\n",
    "    print(\"\\tWorking on {}...\".format(tmp_chr))\n",
    "    seg_tads = tads.loc[tads[\"chr\"] == tmp_chr].copy()\n",
    "    seg_peaks = peaks.loc[peaks[\"chr\"] == tmp_chr].copy()\n",
    "    for i, tmp_tad in seg_tads.iterrows():\n",
    "        tmp_hits = ~((tmp_tad[\"start\"] > seg_peaks[\"end\"]) | \\\n",
    "            (tmp_tad[\"end\"] < seg_peaks[\"start\"]))\n",
    "        tmp_hits = tmp_hits[tmp_hits].index.values\n",
    "        if len(tmp_hits) > 0:\n",
    "            atac_tad_assignments.loc[tmp_hits] = i\n",
    "\n",
    "atac_tad_assignments = atac_tad_assignments[\\\n",
    "    ~atac_tad_assignments.isna()].astype(int)\n",
    "\n",
    "peaks[\"tad\"] = \"\"\n",
    "peaks.loc[atac_tad_assignments.index.values, \"tad\"] = \\\n",
    "    (tads.loc[atac_tad_assignments.values, \"chr\"] + \":\" + \\\n",
    "    tads.loc[atac_tad_assignments.values, \"start\"].astype(str) + \"-\" + \\\n",
    "    tads.loc[atac_tad_assignments.values, \"end\"].astype(str)).values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Assigning TADs to start sites of variably expressed genes...\n",
      "\tWorking on chr10...\n",
      "\tWorking on chr11...\n",
      "\tWorking on chr12...\n",
      "\tWorking on chr19...\n",
      "\tWorking on chrX...\n",
      "\tWorking on chr8...\n",
      "\tWorking on chr13...\n",
      "\tWorking on chr5...\n",
      "\tWorking on chr2...\n",
      "\tWorking on chr6...\n",
      "\tWorking on chr14...\n",
      "\tWorking on chr15...\n",
      "\tWorking on chr18...\n",
      "\tWorking on chr16...\n",
      "\tWorking on chr17...\n",
      "\tWorking on chr4...\n",
      "\tWorking on chr1...\n",
      "\tWorking on chr3...\n",
      "\tWorking on chr7...\n",
      "\tWorking on chr9...\n",
      "\tWorking on chrX_GL456233_random...\n",
      "\tWorking on chrY...\n"
     ]
    }
   ],
   "source": [
    "# Assign TADs to tss start sites\n",
    "print(\"Assigning TADs to start sites of variably expressed genes...\")\n",
    "tss_tad_assignments = pd.Series([np.nan] * tss_loci.shape[0], \n",
    "    index=tss_loci.index.values)\n",
    "for tmp_chr in tads[\"chr\"].unique():\n",
    "    print(\"\\tWorking on {}...\".format(tmp_chr))\n",
    "    seg_tads = tads.loc[tads[\"chr\"] == tmp_chr].copy()\n",
    "    seg_starts = tss_loci.loc[tss_loci[\"chr\"] == tmp_chr].copy()\n",
    "    for i, tmp_tad in seg_tads.iterrows():\n",
    "        tmp_hits = ~((tmp_tad[\"start\"] > seg_starts[\"start\"]) | \\\n",
    "            (tmp_tad[\"end\"] < seg_starts[\"start\"]))\n",
    "        tmp_hits = tmp_hits[tmp_hits].index.values\n",
    "        if len(tmp_hits) > 0:\n",
    "            tss_tad_assignments.loc[tmp_hits] = i\n",
    "\n",
    "tss_tad_assignments = tss_tad_assignments[\\\n",
    "    ~tss_tad_assignments.isna()].astype(int)\n",
    "\n",
    "tss_loci[\"tad\"] = \"\"\n",
    "tss_loci.loc[tss_tad_assignments.index.values, \"tad\"] = \\\n",
    "    (tads.loc[tss_tad_assignments.values, \"chr\"] + \":\" + \\\n",
    "    tads.loc[tss_tad_assignments.values, \"start\"].astype(str) + \"-\" + \\\n",
    "    tads.loc[tss_tad_assignments.values, \"end\"].astype(str)).values\n",
    "\n",
    "tss_loci.index = tss_loci[\"gene\"]\n",
    "# tss_loci = tss_loci.loc[tss_loci[\"gene\"].loc[tss_loci[\"gene\"].isin(\\\n",
    "#     expressed_genes)].values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finding correlations between seqlets and genes in TADs...\n",
      "2093/2093 tads processed\r"
     ]
    }
   ],
   "source": [
    "## Correlate seqlet CS with gene expression within TADs\n",
    "print(\"Finding correlations between seqlets and genes in TADs...\")\n",
    "corr_dict = {}\n",
    "tads_without_seqlets = []\n",
    "unique_tads = tss_loci[\"tad\"].unique()\n",
    "for i,tmp_tad in enumerate(unique_tads):\n",
    "    print(f\"{i+1}/{len(unique_tads)} tads processed\", end=\"\\r\", flush=True)\n",
    "    if tmp_tad != \"\":\n",
    "        seg_gene_anno = tss_loci.loc[tss_loci[\"tad\"] == tmp_tad]\n",
    "        seg_gene_anno = seg_gene_anno.loc[\\\n",
    "            seg_gene_anno[\"gene\"].isin(cite_centroids.columns.values)]\n",
    "        tmp_rna = cite_centroids.loc[\\\n",
    "            dp_scores.columns.values,\n",
    "            seg_gene_anno.index.values] \n",
    "        tmp_peaks = peaks.loc[peaks[\"tad\"] == tmp_tad]\n",
    "        tmp_peaks = (\\\n",
    "            tmp_peaks[\"chr\"] + \":\" + \\\n",
    "            tmp_peaks[\"start\"].astype(str) + \"-\" + \\\n",
    "            tmp_peaks[\"end\"].astype(str)).values\n",
    "        tmp_seqlets = seqlets.loc[seqlets[\"peak\"].isin(tmp_peaks)]\n",
    "        if tmp_seqlets.shape[0] > 0:\n",
    "            tmp_dp_scores = dp_scores.loc[tmp_seqlets.index.values]\n",
    "            r_df = pearson_corr_df_to_df(tmp_rna, tmp_dp_scores.T)\n",
    "            corr_res = r_df.stack().reset_index()\n",
    "            corr_res.columns = [\"Gene\", \"seqlet_idx\", \"r\"]\n",
    "            corr_dict[tmp_tad] = {}\n",
    "            corr_dict[tmp_tad][\"r_df\"] = r_df\n",
    "            corr_dict[tmp_tad][\"sig_conns\"] = corr_res.loc[corr_res[\"r\"] > 0.4]\n",
    "        else:\n",
    "            tads_without_seqlets.append(tmp_tad)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kyle/anaconda3/envs/pyInfinityFlow/lib/python3.8/site-packages/pandas/core/indexing.py:1732: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._setitem_single_block(indexer, value, name)\n",
      "/home/kyle/anaconda3/envs/pyInfinityFlow/lib/python3.8/site-packages/pandas/core/indexing.py:723: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  iloc._setitem_with_indexer(indexer, value, self.name)\n"
     ]
    }
   ],
   "source": [
    "## Change name of tads to prefix with 0s so it is easy to navigate\n",
    "# Build temporary df to store unique tads\n",
    "unique_tads = list(corr_dict.keys())\n",
    "tad_re = re.compile(r'(.*):(.*)-(.*)')\n",
    "unique_tads = pd.DataFrame(\\\n",
    "    [list(tad_re.findall(item)[0]) for item in unique_tads],\n",
    "    columns=[\"chr\", \"start\", \"end\"],\n",
    "    index=unique_tads)\n",
    "\n",
    "# Build list of zfilled ints for TAD positions\n",
    "list_seg_tads = []\n",
    "for tmp_chr in unique_tads[\"chr\"].unique():\n",
    "    seg_tads = unique_tads.loc[unique_tads[\"chr\"] == tmp_chr]\n",
    "    tmp_zfill = seg_tads[\"end\"].apply(len).max()\n",
    "    seg_tads.loc[:,[\"start\", \"end\"]] = seg_tads.loc[:,[\"start\", \"end\"]].applymap(\\\n",
    "        lambda x: x.zfill(tmp_zfill)).values\n",
    "    list_seg_tads.append(seg_tads)\n",
    "    \n",
    "zfilled_tads = pd.concat(list_seg_tads)\n",
    "zfilled_tads[\"zfill\"] = zfilled_tads[\"chr\"] + \"_\" + \\\n",
    "    zfilled_tads[\"start\"] + \"_\" + zfilled_tads[\"end\"]\n",
    "tad_to_zfill = zfilled_tads[[\"zfill\"]]\n",
    "tad_to_zfill.to_csv(\"output/chrombpnet/modisco_merged_results/\"\\\n",
    "    \"fold_0/redo_extract_seqlets/correlate_seqlets_to_gene_expression_all_genes/\"\\\n",
    "    \"tad_to_zfill.csv\", header=False, index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pull out \"significant\" connections to build a separate dataframe\n",
    "sig_conns = []\n",
    "for tmp_tad in corr_dict:\n",
    "    sig_conns.append(corr_dict[tmp_tad][\"sig_conns\"].copy())\n",
    "    \n",
    "sig_conns = pd.concat(sig_conns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Write out the correlations of TAD contained peak to gene connections\n",
    "path_output = \"output/chrombpnet/modisco_merged_results/fold_0/\"\\\n",
    "    \"redo_extract_seqlets/correlate_seqlets_to_gene_expression_all_genes/\"\n",
    "path_test_name = \"seqlet_to_gene_correlation_within_tads\"\n",
    "\n",
    "# Make sure directories exist\n",
    "if not os.path.isdir(path_output):\n",
    "    os.mkdir(path_output)\n",
    "    \n",
    "if not os.path.isdir(os.path.join(path_output, path_test_name)):\n",
    "    os.mkdir(os.path.join(path_output, path_test_name))\n",
    "    \n",
    "if not os.path.isdir(os.path.join(path_output, path_test_name, \"significant_connections\")):\n",
    "    os.mkdir(os.path.join(path_output, path_test_name, \"significant_connections\"))\n",
    "    \n",
    "if not os.path.isdir(os.path.join(path_output, path_test_name, \"r_values\")):\n",
    "    os.mkdir(os.path.join(path_output, path_test_name, \"r_values\"))\n",
    "    \n",
    "\n",
    "\n",
    "# Loop through TADs and write out the correlation results\n",
    "for tmp_tad in unique_tads.index.values:\n",
    "    corr_dict[tmp_tad]['sig_conns'].to_csv(os.path.join(\\\n",
    "        path_output,\n",
    "        path_test_name,\n",
    "        \"significant_connections\",\n",
    "        \"{}_significat_connections.csv\".format(tad_to_zfill[tmp_tad])))\n",
    "    corr_dict[tmp_tad]['r_df'].to_csv(os.path.join(\\\n",
    "        path_output,\n",
    "        path_test_name,\n",
    "        \"r_values\",\n",
    "        \"{}_r_values.csv\".format(tad_to_zfill[tmp_tad])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save annotation files\n",
    "tss_loci.to_csv(\"output/chrombpnet/modisco_merged_results/fold_0/\"\\\n",
    "    \"redo_extract_seqlets/correlate_seqlets_to_gene_expression_all_genes/\"\\\n",
    "    \"gene_loci_annotation.csv\",\n",
    "    header=True, index=True, index_label=\"id\")\n",
    "\n",
    "peaks.to_csv(\"output/chrombpnet/modisco_merged_results/fold_0/\"\\\n",
    "    \"redo_extract_seqlets/correlate_seqlets_to_gene_expression_all_genes/\"\\\n",
    "    \"peak_annotation.csv\",\n",
    "    header=True, index=True, index_label=\"id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_conns.to_csv(\"output/chrombpnet/modisco_merged_results/fold_0/\"\\\n",
    "    \"redo_extract_seqlets/correlate_seqlets_to_gene_expression_all_genes/\"\\\n",
    "    \"sig_conns.csv\",\n",
    "    header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "tad_to_zfill.to_csv(\"output/chrombpnet/modisco_merged_results/fold_0/\"\\\n",
    "    \"redo_extract_seqlets/correlate_seqlets_to_gene_expression_all_genes/\"\\\n",
    "    \"tad_to_zfill.csv\",\n",
    "    header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(30813225, 3)"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_conns.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Read in the results from previous analysis of seqlet to gene expression cor\n",
    "## Gene clusters: hierarchical clusturing across seqlet to gene r values\n",
    "# Read in the gene clusters\n",
    "gene_clusters = pd.read_csv(\\\n",
    "    \"output/chrombpnet/modisco_merged_results/fold_0/redo_extract_seqlets/\"\\\n",
    "    \"seqlet_to_gene_expression_correlation_gene_clusters.csv\",\n",
    "    header=None, index_col=0)\n",
    "gene_clusters.index.name = None\n",
    "## Seqlet clusters: MarkerFinder to hierarchical gene clusters\n",
    "# Read in the seqlet clusters\n",
    "marker_df = pd.read_csv(\\\n",
    "    \"output/chrombpnet/modisco_merged_results/fold_0/redo_extract_seqlets/\"\\\n",
    "    \"seqlet_to_gene_expression_correlation_seqlet_markerfinder.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>seqlet_idx</th>\n",
       "      <th>r</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>1856496</td>\n",
       "      <td>0.441411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>1856497</td>\n",
       "      <td>0.467160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>2458694</td>\n",
       "      <td>0.544898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>2807710</td>\n",
       "      <td>0.544757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>3763246</td>\n",
       "      <td>0.544713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22706190</th>\n",
       "      <td>Hccs</td>\n",
       "      <td>14808102</td>\n",
       "      <td>0.621244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22706191</th>\n",
       "      <td>Hccs</td>\n",
       "      <td>14808105</td>\n",
       "      <td>0.430966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22706192</th>\n",
       "      <td>Hccs</td>\n",
       "      <td>15314097</td>\n",
       "      <td>0.439063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22706193</th>\n",
       "      <td>Hccs</td>\n",
       "      <td>15796481</td>\n",
       "      <td>0.439211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22706194</th>\n",
       "      <td>Hccs</td>\n",
       "      <td>16097276</td>\n",
       "      <td>0.439197</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22706195 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Gene  seqlet_idx         r\n",
       "0         Mrpl15     1856496  0.441411\n",
       "1         Mrpl15     1856497  0.467160\n",
       "2         Mrpl15     2458694  0.544898\n",
       "3         Mrpl15     2807710  0.544757\n",
       "4         Mrpl15     3763246  0.544713\n",
       "...          ...         ...       ...\n",
       "22706190    Hccs    14808102  0.621244\n",
       "22706191    Hccs    14808105  0.430966\n",
       "22706192    Hccs    15314097  0.439063\n",
       "22706193    Hccs    15796481  0.439211\n",
       "22706194    Hccs    16097276  0.439197\n",
       "\n",
       "[22706195 rows x 3 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_conns = pd.read_csv(\"output/chrombpnet/modisco_merged_results/fold_0/\"\\\n",
    "    \"redo_extract_seqlets/correlate_seqlets_to_gene_expression/sig_conns.csv\")\n",
    "\n",
    "sig_conns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9034"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sig_conns[\"Gene\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"Epx\" in sig_conns[\"Gene\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "mfr, mfp = marker_finder(\\\n",
    "    cite_df, \n",
    "    cite_adata.obs.loc[cite_df.index.values, \"lvl3\"].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "mfr = mfr.dropna()\n",
    "stacked_mfr = mfr.stack().reset_index()\n",
    "stacked_mfr.columns = [\"gene\", \"cluster\", \"r\"]\n",
    "stacked_mfr = stacked_mfr.sort_values(by=\"r\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>cluster</th>\n",
       "      <th>r</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>566521</th>\n",
       "      <td>Jchain</td>\n",
       "      <td>plasmaDC</td>\n",
       "      <td>0.951820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173401</th>\n",
       "      <td>Prg2</td>\n",
       "      <td>EoP</td>\n",
       "      <td>0.909323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1764073</th>\n",
       "      <td>Ms4a1</td>\n",
       "      <td>matB</td>\n",
       "      <td>0.882498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>494819</th>\n",
       "      <td>C1qc</td>\n",
       "      <td>Mac-Nr1h3</td>\n",
       "      <td>0.880473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1300051</th>\n",
       "      <td>Epx</td>\n",
       "      <td>EoP</td>\n",
       "      <td>0.874616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408523</th>\n",
       "      <td>Rps20</td>\n",
       "      <td>immNeu-1</td>\n",
       "      <td>-0.547980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>775576</th>\n",
       "      <td>Rps16</td>\n",
       "      <td>immNeu-1</td>\n",
       "      <td>-0.554078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1209532</th>\n",
       "      <td>Rpsa</td>\n",
       "      <td>immNeu-1</td>\n",
       "      <td>-0.555368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>637507</th>\n",
       "      <td>Hmgb1</td>\n",
       "      <td>immNeu-1</td>\n",
       "      <td>-0.575843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44341</th>\n",
       "      <td>Ptma</td>\n",
       "      <td>immNeu-1</td>\n",
       "      <td>-0.674155</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1800291 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene    cluster         r\n",
       "566521   Jchain   plasmaDC  0.951820\n",
       "173401     Prg2        EoP  0.909323\n",
       "1764073   Ms4a1       matB  0.882498\n",
       "494819     C1qc  Mac-Nr1h3  0.880473\n",
       "1300051     Epx        EoP  0.874616\n",
       "...         ...        ...       ...\n",
       "408523    Rps20   immNeu-1 -0.547980\n",
       "775576    Rps16   immNeu-1 -0.554078\n",
       "1209532    Rpsa   immNeu-1 -0.555368\n",
       "637507    Hmgb1   immNeu-1 -0.575843\n",
       "44341      Ptma   immNeu-1 -0.674155\n",
       "\n",
       "[1800291 rows x 3 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stacked_mfr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cite_adata.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in the CITE-seq RNA counts\n",
    "path_cite_data = \"/media/kyle_storage/kyle_ferchen/grimes_lab_main/data/\"\\\n",
    "    \"2021_11_mouse_optimized_cite_seq/processed_files/\"\n",
    "\n",
    "adata_cite = sc.read(os.path.join(\\\n",
    "    path_cite_data, \n",
    "    \"cite_seq_adata_rna_combined.h5ad\"))\n",
    "print(\"Computing CPTT normalized scRNA-seq from CITE-seq...\")\n",
    "adata_cite.X = np.log2((10000 * (adata_cite.X.T / \\\n",
    "    adata_cite.X.sum(axis=1).T).T) + 1)\n",
    "cite_cell_anno = pd.read_csv(os.path.join(\\\n",
    "    path_cite_data, \n",
    "    \"cite_seq_cell_annotations.csv\"))\n",
    "cite_cell_anno.index = cite_cell_anno[\"Cell_Barcode\"].values\n",
    "\n",
    "# Make pseudobulk RNA counts from cite clusters\n",
    "print(\"Computing RNA centroids...\")\n",
    "shared_cells_cite = np.intersect1d(cite_cell_anno.index.values, \n",
    "    adata_cite.obs.index.values)\n",
    "cite_cell_anno = cite_cell_anno.loc[shared_cells_cite]\n",
    "cite_rna = {}\n",
    "for tmp_cluster in cite_cell_anno[\"sctri_cite\"].unique():\n",
    "    print(\"\\t{}...\".format(tmp_cluster))\n",
    "    tmp_barcodes = cite_cell_anno.loc[\\\n",
    "        cite_cell_anno[\"sctri_cite\"] == tmp_cluster].index.values\n",
    "    cite_rna[tmp_cluster] = pd.Series(np.asarray(\\\n",
    "            adata_cite[tmp_barcodes].X.mean(axis=0)).reshape(-1),\n",
    "            index=adata_cite.var.index.values)\n",
    "\n",
    "cite_rna = pd.DataFrame(cite_rna)\n",
    "cite_rna = cite_rna.drop(\"Unknown\", axis=1)\n",
    "map_r7_names = pd.read_csv(os.path.join(\\\n",
    "    path_cite_data, \n",
    "    \"map_r7-v1_to_r7-v2_names.csv\"))\n",
    "map_r7_names_v1_to_v2 = pd.Series(\\\n",
    "    map_r7_names[\"R7_V2\"].values,\n",
    "    index = map_r7_names[\"R7_V1\"].values)\n",
    "\n",
    "cite_rna = cite_rna.rename(map_r7_names_v1_to_v2, axis=1)\n",
    "cite_rna.columns = [i.replace(\"-\", \"_\") for i in cite_rna.columns.values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cite_rna.loc[\"Il5ra\"].sort_values(ascending=False).head(20)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyInfinityFlow",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
