{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:matplotlib.font_manager:Matplotlib is building the font cache; this may take a moment.\n",
      "INFO:matplotlib.font_manager:Failed to extract font properties from /usr/share/fonts/truetype/noto/NotoColorEmoji.ttf: In FT2Font: Can not load face (unknown file format; error code 0x2)\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import re\n",
    "from gimmemotifs.motif import Motif,read_motifs\n",
    "\n",
    "from pyInfinityFlow.InfinityFlow_Utilities import pearson_corr_df_to_df\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rcParams[\"font.family\"] = \"Arial\"\n",
    "import matplotlib\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "\n",
    "os.chdir(\"/media/kyle_storage/kyle_ferchen/grimes_lab_main/analysis/\"\\\n",
    "    \"2023_06_12_tea_seq_atac_processing/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Computing CPTT normalized scRNA-seq from CITE-seq...\n",
      "Computing RNA centroids...\n",
      "\tERP4--Ex...\n",
      "\tMultiLin-2_ML-c9--RNA...\n",
      "\tERP2_Kit-c14--RNA...\n",
      "\tERP2--Ex...\n",
      "\tEbf1+_proB_CD127-c2--RNA...\n",
      "\tproNeu-1_1-4--ADT...\n",
      "\tUnknown...\n",
      "\tpreNeu-2_1-8--ADT...\n",
      "\tcMoP_ML-c15--RNA...\n",
      "\timmNeu_Kit-c3--RNA...\n",
      "\tMP--Ex...\n",
      "\tproNeu-2_Kit-c4--RNA...\n",
      "\tEbf1+_proB--Ex...\n",
      "\tML_ERP1_ML-c4--RNA...\n",
      "\tMDP--Ex...\n",
      "\tpreNeu-3_Kit-c2--RNA...\n",
      "\tERP4_Kit-c13--RNA...\n",
      "\tpreNeu-3_2-3--ADT...\n",
      "\timmNeu_2-5--ADT...\n",
      "\tpreNeu-1_Kit-c5--RNA...\n",
      "\timmNeu_Kit-c1--RNA...\n",
      "\tcMoP--Ex...\n",
      "\tneoHPC_Myc--Ex...\n",
      "\tBcl11b+_preETP--Ex...\n",
      "\tML_Multi-Lin-2_HSC-c7--RNA...\n",
      "\tEbf1+_proB_CD127-c8--RNA...\n",
      "\tIG2-proNeu1_ML-c14--RNA...\n",
      "\tML_Mast--Ex...\n",
      "\tML_cell_cycle_ML-c5--RNA...\n",
      "\tHSC-HPC-Cenpf_HSC-c12--RNA...\n",
      "\tMultiLin-2_ERP_ML-c10--RNA...\n",
      "\tML_MDP_ML-c13--RNA...\n",
      "\tproNeu-1--Ex...\n",
      "\tSiglecH-Ly6C-pre-DC--Ex...\n",
      "\tHSC-HPC-Mki67--Ex...\n",
      "\tPro-B_CD127-c3--RNA...\n",
      "\tprecursor_B_cell_5-2--ADT...\n",
      "\tCLP1_CD127-c5--RNA...\n",
      "\tML_MultiLin-1_ML-c8--RNA...\n",
      "\tETP-A-0-Ccl4--Ex...\n",
      "\tETP-CC-4-Ung--Ex...\n",
      "\teHSC-Pcna_HSC-c9--RNA...\n",
      "\tMPP4-Flt3_HSC-c6--RNA...\n",
      "\tMP_ML-c12--RNA...\n",
      "\tML_Basophil_Kit-c12--RNA...\n",
      "\tB-cell_progenitor--Ex...\n",
      "\tML_ERP1--Ex...\n",
      "\tPlasmaDC--Ex...\n",
      "\timmNeu_21-1--ADT...\n",
      "\tDN4-DP-trans-Hist1h3c_CD127-c12--RNA...\n",
      "\tML_BMCP--Ex...\n",
      "\tILC1/ILC3_NKP_12-1--ADT...\n",
      "\tSiglecH-Ly6C-pre-DC_15-0--ADT...\n",
      "\tML_MEP_ML-c2--RNA...\n",
      "\tML_MKP_16-3--ADT...\n",
      "\tML_MKP--Ex...\n",
      "\tpBplasmaDC--Ex...\n",
      "\tpBplasmaDC_24-8--ADT...\n",
      "\tML_Multi-Lin-1--Ex...\n",
      "\tCLP1--Ex...\n",
      "\tML_BMCP_ML-c7--RNA...\n",
      "\talphaLP--Ex...\n",
      "\tCLP2_CD127-c10--RNA...\n",
      "\tML_Eosinophils_Kit-c11--RNA...\n",
      "\tBcl11b+_preETP_CD127-c6--RNA...\n",
      "\tML_Multi-Lin-1_16-23--ADT...\n",
      "\tILCP--Ex...\n",
      "\tmixed-monocyte_Kit-c10--RNA...\n",
      "\tHSC-HPC-Cenpf--Ex...\n",
      "\tprecursor_B_cell--Ex...\n",
      "\tERP2_4-7--ADT...\n",
      "\tETP-IFN-10-Ifit3--Ex...\n",
      "\tEbf1+_proB_CD127-c13--RNA...\n",
      "\tML_Mast_ML-c1--RNA...\n",
      "\tmixed-monocyte_Kit-c6--RNA...\n",
      "\tneoHSC-Mlt3_HSC-c5--RNA...\n",
      "\tneoHSC-Mlt3_HSC-c2--RNA...\n",
      "\tML_BMCP_7-2--ADT...\n",
      "\tneoHSC-Mlt3--Ex...\n",
      "\tILCP_12-21--ADT...\n",
      "\tpBplasmaDC_CD127-c11--RNA...\n",
      "\teHSC_HSC-c10--RNA...\n",
      "\tEILP_12-4--ADT...\n",
      "\tILC2_12-2--ADT...\n",
      "\tCHILP--Ex...\n",
      "\tneoHSCP-Flt3--Ex...\n",
      "\tML_MEP_ML-c3--RNA...\n",
      "\tMPP4-Flt3_HSC-c3--RNA...\n",
      "\tML_Multi-Lin-2_HSC-c4--RNA...\n",
      "\tMPP4-Flt3--Ex...\n",
      "\tneoHSC-Mlt3_16-1--ADT...\n"
     ]
    }
   ],
   "source": [
    "# Read in the CITE-seq RNA counts\n",
    "path_cite_data = \"/media/kyle_storage/kyle_ferchen/grimes_lab_main/data/\"\\\n",
    "    \"2021_11_mouse_optimized_cite_seq/processed_files/\"\n",
    "\n",
    "adata_cite = sc.read(os.path.join(\\\n",
    "    path_cite_data, \n",
    "    \"cite_seq_adata_rna_combined.h5ad\"))\n",
    "print(\"Computing CPTT normalized scRNA-seq from CITE-seq...\")\n",
    "adata_cite.X = np.log2((10000 * (adata_cite.X.T / \\\n",
    "    adata_cite.X.sum(axis=1).T).T) + 1)\n",
    "cite_cell_anno = pd.read_csv(os.path.join(\\\n",
    "    path_cite_data, \n",
    "    \"cite_seq_cell_annotations.csv\"))\n",
    "cite_cell_anno.index = cite_cell_anno[\"Cell_Barcode\"].values\n",
    "\n",
    "# Make pseudobulk RNA counts from cite clusters\n",
    "print(\"Computing RNA centroids...\")\n",
    "shared_cells_cite = np.intersect1d(cite_cell_anno.index.values, \n",
    "    adata_cite.obs.index.values)\n",
    "cite_cell_anno = cite_cell_anno.loc[shared_cells_cite]\n",
    "cite_rna = {}\n",
    "for tmp_cluster in cite_cell_anno[\"sctri_cite\"].unique():\n",
    "    print(\"\\t{}...\".format(tmp_cluster))\n",
    "    tmp_barcodes = cite_cell_anno.loc[\\\n",
    "        cite_cell_anno[\"sctri_cite\"] == tmp_cluster].index.values\n",
    "    cite_rna[tmp_cluster] = pd.Series(np.asarray(\\\n",
    "            adata_cite[tmp_barcodes].X.mean(axis=0)).reshape(-1),\n",
    "            index=adata_cite.var.index.values)\n",
    "\n",
    "cite_rna = pd.DataFrame(cite_rna)\n",
    "cite_rna = cite_rna.drop(\"Unknown\", axis=1)\n",
    "# map_r7_names = pd.read_csv(os.path.join(\\\n",
    "#     path_cite_data, \n",
    "#     \"map_r7-v1_to_r7-v2_names.csv\"))\n",
    "# map_r7_names_v1_to_v2 = pd.Series(\\\n",
    "#     map_r7_names[\"R7_V2\"].values,\n",
    "#     index = map_r7_names[\"R7_V1\"].values)\n",
    "\n",
    "# cite_rna = cite_rna.rename(map_r7_names_v1_to_v2, axis=1)\n",
    "# cite_rna.columns = [i.replace(\"-\", \"_\") for i in cite_rna.columns.values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in the motifs\n",
    "path_cisbp = \"/media/kyle_storage/kyle_ferchen/grimes_lab_main/reference/\"\\\n",
    "    \"cisbp2/Mus_musculus_2020_06_01_11_53_pm/\"\n",
    "path_pwms = os.path.join(path_cisbp, \"pwms_all_motifs/\")\n",
    "path_cisbp_anno = os.path.join(path_cisbp, \"TF_Information_all_motifs.txt\")\n",
    "\n",
    "cisbp_anno = pd.read_table(path_cisbp_anno)\n",
    "cisbp_anno = cisbp_anno.loc[cisbp_anno[\"Motif_ID\"] != \".\"]\n",
    "\n",
    "cisbp2_motifs = read_motifs(os.path.join(path_cisbp, \"mouse_cisbp2_all_motif_ids_ppm_format.motif\"))\n",
    "cisbp2_names = np.array([str(x) for x in cisbp2_motifs])\n",
    "\n",
    "# Filter cisbp2 annotation to only motifs that have a real entry\n",
    "cisbp_anno = cisbp_anno.loc[cisbp_anno[\"Motif_ID\"].isin([\"_\".join(i.split(\"_\")[:2]) for i in cisbp2_names])]\n",
    "\n",
    "map_id_to_tf = cisbp_anno[[\"Motif_ID\", \"TF_Name\"]].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_cisbp_correlations = \"output/chrombpnet/modisco_merged_results/fold_0/\"\\\n",
    "    \"correlate_modisco_to_cisbp2_motifs/\"\n",
    "\n",
    "corr_files = [i for i in os.listdir(path_to_cisbp_correlations) if \\\n",
    "    i.endswith(\"csv\") and not i.startswith(\".\")]\n",
    "\n",
    "modisco_pattern_re = re.compile(r'(\\w+__pattern_[0-9]+)_\\w+.csv')\n",
    "\n",
    "corr_files = pd.Series(\\\n",
    "    corr_files, \n",
    "    index=[modisco_pattern_re.findall(i)[0] for i in corr_files])\n",
    "\n",
    "pat_corrs = {}\n",
    "for tmp_pat in corr_files.index.values:\n",
    "    pat_corrs[tmp_pat] = pd.read_csv(os.path.join(\\\n",
    "        path_to_cisbp_correlations,\n",
    "        corr_files[tmp_pat]))\n",
    "\n",
    "for tmp_pat in pat_corrs:\n",
    "    pat_corrs[tmp_pat].loc[:,\"motif_pattern\"] = [i.split(\"_\")[-1] \\\n",
    "        for i in pat_corrs[tmp_pat][\"motif\"].values]\n",
    "    pat_corrs[tmp_pat].loc[:,\"motif\"] = [\"_\".join(i.split(\"_\")[:-1]) \\\n",
    "        for i in pat_corrs[tmp_pat][\"motif\"].values]\n",
    "    pat_corrs[tmp_pat] = pat_corrs[tmp_pat].set_index(\"motif\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# new_cluster_order = pd.read_csv(\"/media/kyle_storage/kyle_ferchen/\"\\\n",
    "#     \"grimes_lab_main/data/2021_11_mouse_optimized_cite_seq/processed_files/\"\\\n",
    "#     \"r7_cluster_order_kf_2024_01.csv\")\n",
    "# map_r7_to_lv3 = pd.Series(\\\n",
    "#     new_cluster_order[\"Level 3\"].values,\n",
    "#     index=new_cluster_order[\"Cluster\"].values)\n",
    "# map_lvl3_to_order = pd.Series(\\\n",
    "#     new_cluster_order[\"Order\"].values,\n",
    "#     index=new_cluster_order[\"Level 3\"].values)\n",
    "# map_r7_to_replace_dash = pd.Series(\\\n",
    "#     new_cluster_order[\"Cluster\"].values,\n",
    "#     index=[i.replace(\"-\", \"_\") for i in new_cluster_order[\"Cluster\"].values])\n",
    "# map_new_cluster_to_order = pd.Series(\\\n",
    "#     new_cluster_order[\"Order\"].values,\n",
    "#     index=new_cluster_order[\"Level 3\"].values)\n",
    "# new_cluster_order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Order</th>\n",
       "      <th>R6</th>\n",
       "      <th>R7</th>\n",
       "      <th>lvl4</th>\n",
       "      <th>Color</th>\n",
       "      <th>Group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>neoHSC-Mlt3_HSC-c5--RNA</td>\n",
       "      <td>LT-HSC_Mllt3</td>\n",
       "      <td>qHSC</td>\n",
       "      <td>#040485</td>\n",
       "      <td>HSPC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>neoHSC-Mlt3--Ex</td>\n",
       "      <td>ST-HSC</td>\n",
       "      <td>aHSC</td>\n",
       "      <td>#043075</td>\n",
       "      <td>HSPC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3.0</td>\n",
       "      <td>MPP4-Flt3_HSC-c3--RNA</td>\n",
       "      <td>MPP4-Hlf</td>\n",
       "      <td>HSC-Ly</td>\n",
       "      <td>#0b3d8c</td>\n",
       "      <td>HSPC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4.0</td>\n",
       "      <td>ML_Multi-Lin-2_HSC-c4--RNA</td>\n",
       "      <td>MPP5-Egr1</td>\n",
       "      <td>MPP5-IER</td>\n",
       "      <td>#11489e</td>\n",
       "      <td>HSPC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5.0</td>\n",
       "      <td>ML_Multi-Lin-2_HSC-c7--RNA</td>\n",
       "      <td>MPP5-Flt3</td>\n",
       "      <td>MPP5 Ly-I</td>\n",
       "      <td>#1852ad</td>\n",
       "      <td>HSPC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>83.0</td>\n",
       "      <td>ILCP_12-21--ADT</td>\n",
       "      <td>ILCP-Tox2</td>\n",
       "      <td>ILCP-b</td>\n",
       "      <td>#76c6f5</td>\n",
       "      <td>ILC/NkP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>84.0</td>\n",
       "      <td>ILC2_12-2--ADT</td>\n",
       "      <td>ILC2</td>\n",
       "      <td>ILC2</td>\n",
       "      <td>#5dbdf5</td>\n",
       "      <td>ILC/NkP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>85.0</td>\n",
       "      <td>Bcl11b+_preETP_CD127-c6--RNA</td>\n",
       "      <td>Bcl11b+_preETP_Cd3d</td>\n",
       "      <td>ILC1_3+NKP</td>\n",
       "      <td>#3db6fc</td>\n",
       "      <td>ILC/NkP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>86.0</td>\n",
       "      <td>Bcl11b+_preETP--Ex</td>\n",
       "      <td>Bcl11b+_preETP_Tdrd5</td>\n",
       "      <td>pre-aceNKP</td>\n",
       "      <td>#1cacff</td>\n",
       "      <td>ILC/NkP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>87.0</td>\n",
       "      <td>ILC1/ILC3_NKP_12-1--ADT</td>\n",
       "      <td>ILC1-ILC3-NKP</td>\n",
       "      <td>aceNKP</td>\n",
       "      <td>#00a2ff</td>\n",
       "      <td>ILC/NkP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>90 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Order                            R6                    R7        lvl4  \\\n",
       "0     1.0       neoHSC-Mlt3_HSC-c5--RNA          LT-HSC_Mllt3        qHSC   \n",
       "1     2.0               neoHSC-Mlt3--Ex                ST-HSC        aHSC   \n",
       "2     3.0         MPP4-Flt3_HSC-c3--RNA              MPP4-Hlf      HSC-Ly   \n",
       "3     4.0    ML_Multi-Lin-2_HSC-c4--RNA             MPP5-Egr1    MPP5-IER   \n",
       "4     5.0    ML_Multi-Lin-2_HSC-c7--RNA             MPP5-Flt3   MPP5 Ly-I   \n",
       "..    ...                           ...                   ...         ...   \n",
       "85   83.0               ILCP_12-21--ADT             ILCP-Tox2      ILCP-b   \n",
       "86   84.0                ILC2_12-2--ADT                  ILC2        ILC2   \n",
       "87   85.0  Bcl11b+_preETP_CD127-c6--RNA   Bcl11b+_preETP_Cd3d  ILC1_3+NKP   \n",
       "88   86.0            Bcl11b+_preETP--Ex  Bcl11b+_preETP_Tdrd5  pre-aceNKP   \n",
       "89   87.0       ILC1/ILC3_NKP_12-1--ADT         ILC1-ILC3-NKP      aceNKP   \n",
       "\n",
       "      Color    Group  \n",
       "0   #040485     HSPC  \n",
       "1   #043075     HSPC  \n",
       "2   #0b3d8c     HSPC  \n",
       "3   #11489e     HSPC  \n",
       "4   #1852ad     HSPC  \n",
       "..      ...      ...  \n",
       "85  #76c6f5  ILC/NkP  \n",
       "86  #5dbdf5  ILC/NkP  \n",
       "87  #3db6fc  ILC/NkP  \n",
       "88  #1cacff  ILC/NkP  \n",
       "89  #00a2ff  ILC/NkP  \n",
       "\n",
       "[90 rows x 6 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cluster_anno = pd.read_csv(\"/media/kyle_storage/kyle_ferchen/grimes_lab_main/\"\\\n",
    "    \"data/2021_11_mouse_optimized_cite_seq/processed_files/\"\\\n",
    "    \"cluster_anno_r6_r7_lvl4_and_order.csv\")\n",
    "map_r7_to_lvl4 = pd.Series(\\\n",
    "    cluster_anno[\"lvl4\"].values,\n",
    "    index=cluster_anno[\"R7\"].values)\n",
    "map_lvl4_to_order = pd.Series(\\\n",
    "    cluster_anno[\"Order\"].values,\n",
    "    index=cluster_anno[\"lvl4\"].values).dropna()\n",
    "map_r7_to_replace_dash = pd.Series(\\\n",
    "    cluster_anno[\"R7\"].values,\n",
    "    index=[i.replace(\"-\", \"_\") for i in cluster_anno[\"R7\"].values])\n",
    "map_r6_to_lvl4 = pd.Series(\\\n",
    "    cluster_anno[\"lvl4\"].values,\n",
    "    index=cluster_anno[\"R6\"].values)\n",
    "cluster_anno\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Seqlet annotation\n",
    "seqlets = pd.read_feather(\"output/chrombpnet/modisco_merged_results/fold_0/\"\\\n",
    "    \"redo_extract_seqlets/all_seqlit_hits_above_modisco_min_anno.fea\")\n",
    "\n",
    "# Seqlets\n",
    "dp_scores = pd.read_feather(\"output/chrombpnet/modisco_merged_results/fold_0/\"\\\n",
    "    \"redo_extract_seqlets/all_seqlit_hits_above_modisco_min_dp_scores.fea\")\n",
    "dp_scores = dp_scores.set_index(\"index\").astype(np.float32)\n",
    "dp_scores = dp_scores.rename(map_r7_to_replace_dash.to_dict(), axis=1)\n",
    "dp_scores = dp_scores.rename(map_r7_to_lvl4.to_dict(), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ErP-5</th>\n",
       "      <th>ML-MDP</th>\n",
       "      <th>ErP-3</th>\n",
       "      <th>ErP-2</th>\n",
       "      <th>preB-kappa</th>\n",
       "      <th>proNeu-1b</th>\n",
       "      <th>preNeu-2</th>\n",
       "      <th>cMoP-a</th>\n",
       "      <th>immNeu-1</th>\n",
       "      <th>MP</th>\n",
       "      <th>...</th>\n",
       "      <th>MPP1-MKP</th>\n",
       "      <th>CHILP</th>\n",
       "      <th>ILC2</th>\n",
       "      <th>EILP</th>\n",
       "      <th>MPP3-IER</th>\n",
       "      <th>MEP</th>\n",
       "      <th>HSC-Ly</th>\n",
       "      <th>MPP5-IER</th>\n",
       "      <th>MPP4-TCA</th>\n",
       "      <th>HSC-Mac-2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Xkr4</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm1992</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm37381</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rp1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000992</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000859</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rp1.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC168977.1</th>\n",
       "      <td>0.000908</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000177</td>\n",
       "      <td>0.000676</td>\n",
       "      <td>0.001885</td>\n",
       "      <td>0.000388</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001044</td>\n",
       "      <td>0.001121</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001185</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002227</td>\n",
       "      <td>0.001502</td>\n",
       "      <td>0.002648</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PISD</th>\n",
       "      <td>0.130867</td>\n",
       "      <td>0.487255</td>\n",
       "      <td>0.218956</td>\n",
       "      <td>0.291408</td>\n",
       "      <td>0.645974</td>\n",
       "      <td>1.015920</td>\n",
       "      <td>0.562707</td>\n",
       "      <td>0.496978</td>\n",
       "      <td>0.203916</td>\n",
       "      <td>0.693699</td>\n",
       "      <td>...</td>\n",
       "      <td>0.852288</td>\n",
       "      <td>0.973890</td>\n",
       "      <td>0.763850</td>\n",
       "      <td>0.577835</td>\n",
       "      <td>1.094302</td>\n",
       "      <td>0.977848</td>\n",
       "      <td>1.630157</td>\n",
       "      <td>1.226136</td>\n",
       "      <td>1.565246</td>\n",
       "      <td>1.271523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DHRSX</th>\n",
       "      <td>0.406684</td>\n",
       "      <td>0.332551</td>\n",
       "      <td>0.404806</td>\n",
       "      <td>0.390642</td>\n",
       "      <td>0.155894</td>\n",
       "      <td>0.376706</td>\n",
       "      <td>0.215978</td>\n",
       "      <td>0.418623</td>\n",
       "      <td>0.081948</td>\n",
       "      <td>0.433763</td>\n",
       "      <td>...</td>\n",
       "      <td>0.219337</td>\n",
       "      <td>0.315286</td>\n",
       "      <td>0.215548</td>\n",
       "      <td>0.202145</td>\n",
       "      <td>0.183721</td>\n",
       "      <td>0.200725</td>\n",
       "      <td>0.171802</td>\n",
       "      <td>0.203645</td>\n",
       "      <td>0.174204</td>\n",
       "      <td>0.206222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Vmn2r122</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAAA01147332.1</th>\n",
       "      <td>0.006353</td>\n",
       "      <td>0.025774</td>\n",
       "      <td>0.006254</td>\n",
       "      <td>0.006947</td>\n",
       "      <td>0.012995</td>\n",
       "      <td>0.012221</td>\n",
       "      <td>0.007744</td>\n",
       "      <td>0.020161</td>\n",
       "      <td>0.001186</td>\n",
       "      <td>0.020004</td>\n",
       "      <td>...</td>\n",
       "      <td>0.017395</td>\n",
       "      <td>0.062695</td>\n",
       "      <td>0.008949</td>\n",
       "      <td>0.015922</td>\n",
       "      <td>0.023223</td>\n",
       "      <td>0.012274</td>\n",
       "      <td>0.010137</td>\n",
       "      <td>0.016466</td>\n",
       "      <td>0.018917</td>\n",
       "      <td>0.013758</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28692 rows × 90 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   ErP-5    ML-MDP     ErP-3     ErP-2  preB-kappa  proNeu-1b  \\\n",
       "Xkr4            0.000000  0.000000  0.000000  0.000000    0.000000   0.000000   \n",
       "Gm1992          0.000000  0.000000  0.000000  0.000000    0.000000   0.000000   \n",
       "Gm37381         0.000000  0.000000  0.000000  0.000000    0.000000   0.000000   \n",
       "Rp1             0.000000  0.000000  0.000000  0.000000    0.000000   0.000000   \n",
       "Rp1.1           0.000000  0.000000  0.000000  0.000000    0.000000   0.000000   \n",
       "...                  ...       ...       ...       ...         ...        ...   \n",
       "AC168977.1      0.000908  0.000000  0.000177  0.000676    0.001885   0.000388   \n",
       "PISD            0.130867  0.487255  0.218956  0.291408    0.645974   1.015920   \n",
       "DHRSX           0.406684  0.332551  0.404806  0.390642    0.155894   0.376706   \n",
       "Vmn2r122        0.000000  0.000000  0.000000  0.000000    0.000000   0.000000   \n",
       "CAAA01147332.1  0.006353  0.025774  0.006254  0.006947    0.012995   0.012221   \n",
       "\n",
       "                preNeu-2    cMoP-a  immNeu-1        MP  ...  MPP1-MKP  \\\n",
       "Xkr4            0.000000  0.000000  0.000000  0.000000  ...  0.000000   \n",
       "Gm1992          0.000000  0.000000  0.000000  0.000000  ...  0.000000   \n",
       "Gm37381         0.000000  0.000000  0.000000  0.000000  ...  0.000000   \n",
       "Rp1             0.000000  0.000992  0.000000  0.000859  ...  0.000000   \n",
       "Rp1.1           0.000000  0.000000  0.000000  0.000000  ...  0.000000   \n",
       "...                  ...       ...       ...       ...  ...       ...   \n",
       "AC168977.1      0.000000  0.001044  0.001121  0.000000  ...  0.001185   \n",
       "PISD            0.562707  0.496978  0.203916  0.693699  ...  0.852288   \n",
       "DHRSX           0.215978  0.418623  0.081948  0.433763  ...  0.219337   \n",
       "Vmn2r122        0.000000  0.000000  0.000000  0.000000  ...  0.000000   \n",
       "CAAA01147332.1  0.007744  0.020161  0.001186  0.020004  ...  0.017395   \n",
       "\n",
       "                   CHILP      ILC2      EILP  MPP3-IER       MEP    HSC-Ly  \\\n",
       "Xkr4            0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "Gm1992          0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "Gm37381         0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "Rp1             0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "Rp1.1           0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "...                  ...       ...       ...       ...       ...       ...   \n",
       "AC168977.1      0.000000  0.000000  0.000000  0.000000  0.000000  0.002227   \n",
       "PISD            0.973890  0.763850  0.577835  1.094302  0.977848  1.630157   \n",
       "DHRSX           0.315286  0.215548  0.202145  0.183721  0.200725  0.171802   \n",
       "Vmn2r122        0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "CAAA01147332.1  0.062695  0.008949  0.015922  0.023223  0.012274  0.010137   \n",
       "\n",
       "                MPP5-IER  MPP4-TCA  HSC-Mac-2  \n",
       "Xkr4            0.000000  0.000000   0.000000  \n",
       "Gm1992          0.000000  0.000000   0.000000  \n",
       "Gm37381         0.000000  0.000000   0.000000  \n",
       "Rp1             0.000000  0.000000   0.000000  \n",
       "Rp1.1           0.000000  0.000000   0.000000  \n",
       "...                  ...       ...        ...  \n",
       "AC168977.1      0.001502  0.002648   0.000000  \n",
       "PISD            1.226136  1.565246   1.271523  \n",
       "DHRSX           0.203645  0.174204   0.206222  \n",
       "Vmn2r122        0.000000  0.000000   0.000000  \n",
       "CAAA01147332.1  0.016466  0.018917   0.013758  \n",
       "\n",
       "[28692 rows x 90 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Rename cite-rna to the lvl4 annotations\n",
    "cite_rna = cite_rna.rename(map_r6_to_lvl4.to_dict(), axis=1)\n",
    "cite_rna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>seqlet_idx</th>\n",
       "      <th>r</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>peak</th>\n",
       "      <th>score</th>\n",
       "      <th>pos</th>\n",
       "      <th>strand</th>\n",
       "      <th>pattern</th>\n",
       "      <th>in_modisco</th>\n",
       "      <th>seq_name</th>\n",
       "      <th>dp_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>2458694</td>\n",
       "      <td>0.547196</td>\n",
       "      <td>chr1</td>\n",
       "      <td>4540571</td>\n",
       "      <td>4540601</td>\n",
       "      <td>chr1:4540111-4541111</td>\n",
       "      <td>8.932149</td>\n",
       "      <td>461</td>\n",
       "      <td>-</td>\n",
       "      <td>ERP1__pattern_0</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1:4540571-4540601</td>\n",
       "      <td>0.481389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>2807710</td>\n",
       "      <td>0.547052</td>\n",
       "      <td>chr1</td>\n",
       "      <td>4540571</td>\n",
       "      <td>4540601</td>\n",
       "      <td>chr1:4540111-4541111</td>\n",
       "      <td>8.829016</td>\n",
       "      <td>461</td>\n",
       "      <td>-</td>\n",
       "      <td>ERP2__pattern_0</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1:4540571-4540601</td>\n",
       "      <td>0.405852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>3763246</td>\n",
       "      <td>0.547004</td>\n",
       "      <td>chr1</td>\n",
       "      <td>4540571</td>\n",
       "      <td>4540601</td>\n",
       "      <td>chr1:4540111-4541111</td>\n",
       "      <td>9.052924</td>\n",
       "      <td>461</td>\n",
       "      <td>+</td>\n",
       "      <td>HSCP_ERP1__pattern_1</td>\n",
       "      <td>False</td>\n",
       "      <td>chr1:4540571-4540601</td>\n",
       "      <td>0.083670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>9202943</td>\n",
       "      <td>0.547779</td>\n",
       "      <td>chr1</td>\n",
       "      <td>4540571</td>\n",
       "      <td>4540601</td>\n",
       "      <td>chr1:4540111-4541111</td>\n",
       "      <td>8.155797</td>\n",
       "      <td>461</td>\n",
       "      <td>+</td>\n",
       "      <td>MEP__pattern_3</td>\n",
       "      <td>False</td>\n",
       "      <td>chr1:4540571-4540601</td>\n",
       "      <td>0.157136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>9906506</td>\n",
       "      <td>0.547904</td>\n",
       "      <td>chr1</td>\n",
       "      <td>4540570</td>\n",
       "      <td>4540600</td>\n",
       "      <td>chr1:4540111-4541111</td>\n",
       "      <td>8.989227</td>\n",
       "      <td>460</td>\n",
       "      <td>+</td>\n",
       "      <td>MKP__pattern_3</td>\n",
       "      <td>False</td>\n",
       "      <td>chr1:4540570-4540600</td>\n",
       "      <td>0.196055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30813220</th>\n",
       "      <td>Hccs</td>\n",
       "      <td>14808102</td>\n",
       "      <td>0.618932</td>\n",
       "      <td>chrX</td>\n",
       "      <td>169062729</td>\n",
       "      <td>169062759</td>\n",
       "      <td>chrX:169062299-169063299</td>\n",
       "      <td>7.126749</td>\n",
       "      <td>431</td>\n",
       "      <td>+</td>\n",
       "      <td>pre_MultiLin_1__pattern_11</td>\n",
       "      <td>False</td>\n",
       "      <td>chrX:169062729-169062759</td>\n",
       "      <td>0.142939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30813221</th>\n",
       "      <td>Hccs</td>\n",
       "      <td>14808105</td>\n",
       "      <td>0.434179</td>\n",
       "      <td>chrX</td>\n",
       "      <td>169094790</td>\n",
       "      <td>169094820</td>\n",
       "      <td>chrX:169094308-169095308</td>\n",
       "      <td>10.263774</td>\n",
       "      <td>483</td>\n",
       "      <td>-</td>\n",
       "      <td>pre_MultiLin_1__pattern_4</td>\n",
       "      <td>False</td>\n",
       "      <td>chrX:169094790-169094820</td>\n",
       "      <td>0.406595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30813222</th>\n",
       "      <td>Hccs</td>\n",
       "      <td>15314097</td>\n",
       "      <td>0.441781</td>\n",
       "      <td>chrX</td>\n",
       "      <td>169094790</td>\n",
       "      <td>169094820</td>\n",
       "      <td>chrX:169094308-169095308</td>\n",
       "      <td>9.768803</td>\n",
       "      <td>483</td>\n",
       "      <td>+</td>\n",
       "      <td>pre_MultiLin_2__pattern_5</td>\n",
       "      <td>False</td>\n",
       "      <td>chrX:169094790-169094820</td>\n",
       "      <td>0.165522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30813223</th>\n",
       "      <td>Hccs</td>\n",
       "      <td>15796481</td>\n",
       "      <td>0.441943</td>\n",
       "      <td>chrX</td>\n",
       "      <td>169094786</td>\n",
       "      <td>169094816</td>\n",
       "      <td>chrX:169094308-169095308</td>\n",
       "      <td>10.079901</td>\n",
       "      <td>479</td>\n",
       "      <td>-</td>\n",
       "      <td>proNeu_1__pattern_7</td>\n",
       "      <td>False</td>\n",
       "      <td>chrX:169094786-169094816</td>\n",
       "      <td>0.222236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30813224</th>\n",
       "      <td>Hccs</td>\n",
       "      <td>16097276</td>\n",
       "      <td>0.441947</td>\n",
       "      <td>chrX</td>\n",
       "      <td>169094792</td>\n",
       "      <td>169094822</td>\n",
       "      <td>chrX:169094308-169095308</td>\n",
       "      <td>9.944236</td>\n",
       "      <td>485</td>\n",
       "      <td>+</td>\n",
       "      <td>ST_HSC__pattern_6</td>\n",
       "      <td>False</td>\n",
       "      <td>chrX:169094792-169094822</td>\n",
       "      <td>0.218748</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30695180 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Gene  seqlet_idx         r   chr      start        end  \\\n",
       "2         Mrpl15     2458694  0.547196  chr1    4540571    4540601   \n",
       "3         Mrpl15     2807710  0.547052  chr1    4540571    4540601   \n",
       "4         Mrpl15     3763246  0.547004  chr1    4540571    4540601   \n",
       "5         Mrpl15     9202943  0.547779  chr1    4540571    4540601   \n",
       "6         Mrpl15     9906506  0.547904  chr1    4540570    4540600   \n",
       "...          ...         ...       ...   ...        ...        ...   \n",
       "30813220    Hccs    14808102  0.618932  chrX  169062729  169062759   \n",
       "30813221    Hccs    14808105  0.434179  chrX  169094790  169094820   \n",
       "30813222    Hccs    15314097  0.441781  chrX  169094790  169094820   \n",
       "30813223    Hccs    15796481  0.441943  chrX  169094786  169094816   \n",
       "30813224    Hccs    16097276  0.441947  chrX  169094792  169094822   \n",
       "\n",
       "                              peak      score  pos strand  \\\n",
       "2             chr1:4540111-4541111   8.932149  461      -   \n",
       "3             chr1:4540111-4541111   8.829016  461      -   \n",
       "4             chr1:4540111-4541111   9.052924  461      +   \n",
       "5             chr1:4540111-4541111   8.155797  461      +   \n",
       "6             chr1:4540111-4541111   8.989227  460      +   \n",
       "...                            ...        ...  ...    ...   \n",
       "30813220  chrX:169062299-169063299   7.126749  431      +   \n",
       "30813221  chrX:169094308-169095308  10.263774  483      -   \n",
       "30813222  chrX:169094308-169095308   9.768803  483      +   \n",
       "30813223  chrX:169094308-169095308  10.079901  479      -   \n",
       "30813224  chrX:169094308-169095308   9.944236  485      +   \n",
       "\n",
       "                             pattern  in_modisco                  seq_name  \\\n",
       "2                    ERP1__pattern_0        True      chr1:4540571-4540601   \n",
       "3                    ERP2__pattern_0        True      chr1:4540571-4540601   \n",
       "4               HSCP_ERP1__pattern_1       False      chr1:4540571-4540601   \n",
       "5                     MEP__pattern_3       False      chr1:4540571-4540601   \n",
       "6                     MKP__pattern_3       False      chr1:4540570-4540600   \n",
       "...                              ...         ...                       ...   \n",
       "30813220  pre_MultiLin_1__pattern_11       False  chrX:169062729-169062759   \n",
       "30813221   pre_MultiLin_1__pattern_4       False  chrX:169094790-169094820   \n",
       "30813222   pre_MultiLin_2__pattern_5       False  chrX:169094790-169094820   \n",
       "30813223         proNeu_1__pattern_7       False  chrX:169094786-169094816   \n",
       "30813224           ST_HSC__pattern_6       False  chrX:169094792-169094822   \n",
       "\n",
       "          dp_score  \n",
       "2         0.481389  \n",
       "3         0.405852  \n",
       "4         0.083670  \n",
       "5         0.157136  \n",
       "6         0.196055  \n",
       "...            ...  \n",
       "30813220  0.142939  \n",
       "30813221  0.406595  \n",
       "30813222  0.165522  \n",
       "30813223  0.222236  \n",
       "30813224  0.218748  \n",
       "\n",
       "[30695180 rows x 14 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_to_seqlets = pd.read_csv(\"output/chrombpnet/modisco_merged_results/\"\\\n",
    "    \"fold_0/redo_extract_seqlets/\"\\\n",
    "    \"correlate_seqlets_to_gene_expression_all_genes/sig_conns.csv\")\n",
    "gene_to_seqlets = pd.concat([\\\n",
    "    gene_to_seqlets,\n",
    "    seqlets.loc[gene_to_seqlets[\"seqlet_idx\"].values].reset_index(drop=True)], \n",
    "        axis=1)\n",
    "gene_to_seqlets = gene_to_seqlets.loc[\\\n",
    "    ~gene_to_seqlets[\"pattern\"].str.contains(\"__neg_pattern_\")]\n",
    "gene_to_seqlets[\"pattern\"] = gene_to_seqlets[\"pattern\"].str.replace(\\\n",
    "    \"__pos_pattern_\", \"__pattern_\")\n",
    "gene_to_seqlets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mrpl15</th>\n",
       "      <th>Lypla1</th>\n",
       "      <th>Tcea1</th>\n",
       "      <th>Atp6v1h</th>\n",
       "      <th>4732440D04Rik</th>\n",
       "      <th>Rb1cc1</th>\n",
       "      <th>St18</th>\n",
       "      <th>Pcmtd1</th>\n",
       "      <th>Rrs1</th>\n",
       "      <th>Adhfe1</th>\n",
       "      <th>...</th>\n",
       "      <th>Tceanc</th>\n",
       "      <th>Tmsb4x</th>\n",
       "      <th>Tlr8</th>\n",
       "      <th>Tlr7</th>\n",
       "      <th>Prps2</th>\n",
       "      <th>Msl3</th>\n",
       "      <th>Arhgap6</th>\n",
       "      <th>Gm15261</th>\n",
       "      <th>Amelx</th>\n",
       "      <th>Hccs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AC168977.1</th>\n",
       "      <td>-0.207345</td>\n",
       "      <td>0.214908</td>\n",
       "      <td>0.310641</td>\n",
       "      <td>0.447806</td>\n",
       "      <td>0.100249</td>\n",
       "      <td>0.101964</td>\n",
       "      <td>-0.096922</td>\n",
       "      <td>0.254469</td>\n",
       "      <td>-0.264164</td>\n",
       "      <td>-0.008615</td>\n",
       "      <td>...</td>\n",
       "      <td>0.118304</td>\n",
       "      <td>0.223562</td>\n",
       "      <td>-0.190263</td>\n",
       "      <td>0.525339</td>\n",
       "      <td>0.252252</td>\n",
       "      <td>0.071206</td>\n",
       "      <td>0.532635</td>\n",
       "      <td>0.484260</td>\n",
       "      <td>0.096185</td>\n",
       "      <td>0.271224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ahctf1</th>\n",
       "      <td>0.351291</td>\n",
       "      <td>-0.283189</td>\n",
       "      <td>0.721642</td>\n",
       "      <td>-0.395829</td>\n",
       "      <td>0.042972</td>\n",
       "      <td>-0.182254</td>\n",
       "      <td>-0.124383</td>\n",
       "      <td>-0.044344</td>\n",
       "      <td>0.245067</td>\n",
       "      <td>-0.047850</td>\n",
       "      <td>...</td>\n",
       "      <td>0.129361</td>\n",
       "      <td>-0.714817</td>\n",
       "      <td>-0.255081</td>\n",
       "      <td>-0.235827</td>\n",
       "      <td>-0.117408</td>\n",
       "      <td>0.580536</td>\n",
       "      <td>0.132948</td>\n",
       "      <td>0.077102</td>\n",
       "      <td>0.121676</td>\n",
       "      <td>0.456680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ahr</th>\n",
       "      <td>-0.740351</td>\n",
       "      <td>-0.110981</td>\n",
       "      <td>-0.349024</td>\n",
       "      <td>0.424515</td>\n",
       "      <td>0.227347</td>\n",
       "      <td>0.653538</td>\n",
       "      <td>0.412167</td>\n",
       "      <td>0.820725</td>\n",
       "      <td>-0.752553</td>\n",
       "      <td>0.123100</td>\n",
       "      <td>...</td>\n",
       "      <td>0.510348</td>\n",
       "      <td>0.125170</td>\n",
       "      <td>-0.225358</td>\n",
       "      <td>0.090179</td>\n",
       "      <td>-0.244987</td>\n",
       "      <td>-0.515346</td>\n",
       "      <td>0.368094</td>\n",
       "      <td>0.170119</td>\n",
       "      <td>-0.042461</td>\n",
       "      <td>-0.491934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Aire</th>\n",
       "      <td>-0.663790</td>\n",
       "      <td>0.029738</td>\n",
       "      <td>-0.197959</td>\n",
       "      <td>0.366057</td>\n",
       "      <td>0.538739</td>\n",
       "      <td>0.709377</td>\n",
       "      <td>0.207697</td>\n",
       "      <td>0.789649</td>\n",
       "      <td>-0.731934</td>\n",
       "      <td>-0.235657</td>\n",
       "      <td>...</td>\n",
       "      <td>0.640141</td>\n",
       "      <td>-0.040788</td>\n",
       "      <td>-0.218868</td>\n",
       "      <td>0.102753</td>\n",
       "      <td>-0.377064</td>\n",
       "      <td>-0.353798</td>\n",
       "      <td>0.356167</td>\n",
       "      <td>0.160371</td>\n",
       "      <td>-0.075698</td>\n",
       "      <td>-0.337738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Alx1</th>\n",
       "      <td>-0.138140</td>\n",
       "      <td>0.128608</td>\n",
       "      <td>0.109710</td>\n",
       "      <td>-0.015732</td>\n",
       "      <td>-0.026644</td>\n",
       "      <td>0.148997</td>\n",
       "      <td>-0.102979</td>\n",
       "      <td>0.141999</td>\n",
       "      <td>-0.171809</td>\n",
       "      <td>-0.118312</td>\n",
       "      <td>...</td>\n",
       "      <td>0.260712</td>\n",
       "      <td>-0.007479</td>\n",
       "      <td>-0.064929</td>\n",
       "      <td>-0.067750</td>\n",
       "      <td>-0.030915</td>\n",
       "      <td>0.259122</td>\n",
       "      <td>0.122319</td>\n",
       "      <td>0.134667</td>\n",
       "      <td>0.762830</td>\n",
       "      <td>0.038794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zscan20</th>\n",
       "      <td>-0.739673</td>\n",
       "      <td>0.189856</td>\n",
       "      <td>-0.152174</td>\n",
       "      <td>0.428758</td>\n",
       "      <td>0.495468</td>\n",
       "      <td>0.751963</td>\n",
       "      <td>0.489203</td>\n",
       "      <td>0.829072</td>\n",
       "      <td>-0.798257</td>\n",
       "      <td>-0.025336</td>\n",
       "      <td>...</td>\n",
       "      <td>0.796856</td>\n",
       "      <td>0.099543</td>\n",
       "      <td>-0.190500</td>\n",
       "      <td>0.121783</td>\n",
       "      <td>-0.270472</td>\n",
       "      <td>-0.274467</td>\n",
       "      <td>0.466787</td>\n",
       "      <td>0.231449</td>\n",
       "      <td>0.197603</td>\n",
       "      <td>-0.317767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zscan26</th>\n",
       "      <td>-0.806257</td>\n",
       "      <td>-0.116374</td>\n",
       "      <td>-0.428609</td>\n",
       "      <td>0.402260</td>\n",
       "      <td>0.430397</td>\n",
       "      <td>0.836742</td>\n",
       "      <td>0.543816</td>\n",
       "      <td>0.901533</td>\n",
       "      <td>-0.828382</td>\n",
       "      <td>-0.029879</td>\n",
       "      <td>...</td>\n",
       "      <td>0.711038</td>\n",
       "      <td>0.006284</td>\n",
       "      <td>-0.157231</td>\n",
       "      <td>0.022427</td>\n",
       "      <td>-0.498710</td>\n",
       "      <td>-0.531891</td>\n",
       "      <td>0.301573</td>\n",
       "      <td>0.103166</td>\n",
       "      <td>-0.093738</td>\n",
       "      <td>-0.532424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zscan29</th>\n",
       "      <td>-0.705735</td>\n",
       "      <td>0.153951</td>\n",
       "      <td>-0.426196</td>\n",
       "      <td>0.534282</td>\n",
       "      <td>0.415066</td>\n",
       "      <td>0.759371</td>\n",
       "      <td>0.145234</td>\n",
       "      <td>0.714268</td>\n",
       "      <td>-0.626279</td>\n",
       "      <td>-0.002207</td>\n",
       "      <td>...</td>\n",
       "      <td>0.509478</td>\n",
       "      <td>0.241187</td>\n",
       "      <td>-0.185488</td>\n",
       "      <td>-0.156000</td>\n",
       "      <td>-0.562612</td>\n",
       "      <td>-0.432589</td>\n",
       "      <td>0.207287</td>\n",
       "      <td>0.076315</td>\n",
       "      <td>-0.175622</td>\n",
       "      <td>-0.352948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zscan4d</th>\n",
       "      <td>0.000730</td>\n",
       "      <td>0.228454</td>\n",
       "      <td>-0.148824</td>\n",
       "      <td>0.068063</td>\n",
       "      <td>0.229488</td>\n",
       "      <td>0.124778</td>\n",
       "      <td>0.482092</td>\n",
       "      <td>-0.087063</td>\n",
       "      <td>-0.078548</td>\n",
       "      <td>-0.118312</td>\n",
       "      <td>...</td>\n",
       "      <td>0.048984</td>\n",
       "      <td>0.049833</td>\n",
       "      <td>-0.064929</td>\n",
       "      <td>0.054471</td>\n",
       "      <td>0.126177</td>\n",
       "      <td>0.197613</td>\n",
       "      <td>-0.056416</td>\n",
       "      <td>-0.018415</td>\n",
       "      <td>-0.046150</td>\n",
       "      <td>0.067401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zzz3</th>\n",
       "      <td>-0.326547</td>\n",
       "      <td>0.228016</td>\n",
       "      <td>0.614236</td>\n",
       "      <td>0.198179</td>\n",
       "      <td>0.466278</td>\n",
       "      <td>0.291507</td>\n",
       "      <td>0.015399</td>\n",
       "      <td>0.448506</td>\n",
       "      <td>-0.282018</td>\n",
       "      <td>-0.058890</td>\n",
       "      <td>...</td>\n",
       "      <td>0.537148</td>\n",
       "      <td>-0.295525</td>\n",
       "      <td>-0.462171</td>\n",
       "      <td>-0.061166</td>\n",
       "      <td>-0.059481</td>\n",
       "      <td>0.277064</td>\n",
       "      <td>0.677568</td>\n",
       "      <td>0.544409</td>\n",
       "      <td>0.201029</td>\n",
       "      <td>0.354761</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>727 rows × 14137 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Mrpl15    Lypla1     Tcea1   Atp6v1h  4732440D04Rik    Rb1cc1  \\\n",
       "AC168977.1 -0.207345  0.214908  0.310641  0.447806       0.100249  0.101964   \n",
       "Ahctf1      0.351291 -0.283189  0.721642 -0.395829       0.042972 -0.182254   \n",
       "Ahr        -0.740351 -0.110981 -0.349024  0.424515       0.227347  0.653538   \n",
       "Aire       -0.663790  0.029738 -0.197959  0.366057       0.538739  0.709377   \n",
       "Alx1       -0.138140  0.128608  0.109710 -0.015732      -0.026644  0.148997   \n",
       "...              ...       ...       ...       ...            ...       ...   \n",
       "Zscan20    -0.739673  0.189856 -0.152174  0.428758       0.495468  0.751963   \n",
       "Zscan26    -0.806257 -0.116374 -0.428609  0.402260       0.430397  0.836742   \n",
       "Zscan29    -0.705735  0.153951 -0.426196  0.534282       0.415066  0.759371   \n",
       "Zscan4d     0.000730  0.228454 -0.148824  0.068063       0.229488  0.124778   \n",
       "Zzz3       -0.326547  0.228016  0.614236  0.198179       0.466278  0.291507   \n",
       "\n",
       "                St18    Pcmtd1      Rrs1    Adhfe1  ...    Tceanc    Tmsb4x  \\\n",
       "AC168977.1 -0.096922  0.254469 -0.264164 -0.008615  ...  0.118304  0.223562   \n",
       "Ahctf1     -0.124383 -0.044344  0.245067 -0.047850  ...  0.129361 -0.714817   \n",
       "Ahr         0.412167  0.820725 -0.752553  0.123100  ...  0.510348  0.125170   \n",
       "Aire        0.207697  0.789649 -0.731934 -0.235657  ...  0.640141 -0.040788   \n",
       "Alx1       -0.102979  0.141999 -0.171809 -0.118312  ...  0.260712 -0.007479   \n",
       "...              ...       ...       ...       ...  ...       ...       ...   \n",
       "Zscan20     0.489203  0.829072 -0.798257 -0.025336  ...  0.796856  0.099543   \n",
       "Zscan26     0.543816  0.901533 -0.828382 -0.029879  ...  0.711038  0.006284   \n",
       "Zscan29     0.145234  0.714268 -0.626279 -0.002207  ...  0.509478  0.241187   \n",
       "Zscan4d     0.482092 -0.087063 -0.078548 -0.118312  ...  0.048984  0.049833   \n",
       "Zzz3        0.015399  0.448506 -0.282018 -0.058890  ...  0.537148 -0.295525   \n",
       "\n",
       "                Tlr8      Tlr7     Prps2      Msl3   Arhgap6   Gm15261  \\\n",
       "AC168977.1 -0.190263  0.525339  0.252252  0.071206  0.532635  0.484260   \n",
       "Ahctf1     -0.255081 -0.235827 -0.117408  0.580536  0.132948  0.077102   \n",
       "Ahr        -0.225358  0.090179 -0.244987 -0.515346  0.368094  0.170119   \n",
       "Aire       -0.218868  0.102753 -0.377064 -0.353798  0.356167  0.160371   \n",
       "Alx1       -0.064929 -0.067750 -0.030915  0.259122  0.122319  0.134667   \n",
       "...              ...       ...       ...       ...       ...       ...   \n",
       "Zscan20    -0.190500  0.121783 -0.270472 -0.274467  0.466787  0.231449   \n",
       "Zscan26    -0.157231  0.022427 -0.498710 -0.531891  0.301573  0.103166   \n",
       "Zscan29    -0.185488 -0.156000 -0.562612 -0.432589  0.207287  0.076315   \n",
       "Zscan4d    -0.064929  0.054471  0.126177  0.197613 -0.056416 -0.018415   \n",
       "Zzz3       -0.462171 -0.061166 -0.059481  0.277064  0.677568  0.544409   \n",
       "\n",
       "               Amelx      Hccs  \n",
       "AC168977.1  0.096185  0.271224  \n",
       "Ahctf1      0.121676  0.456680  \n",
       "Ahr        -0.042461 -0.491934  \n",
       "Aire       -0.075698 -0.337738  \n",
       "Alx1        0.762830  0.038794  \n",
       "...              ...       ...  \n",
       "Zscan20     0.197603 -0.317767  \n",
       "Zscan26    -0.093738 -0.532424  \n",
       "Zscan29    -0.175622 -0.352948  \n",
       "Zscan4d    -0.046150  0.067401  \n",
       "Zzz3        0.201029  0.354761  \n",
       "\n",
       "[727 rows x 14137 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Build TF to gene correlation matrix (pseudobulk CITE-seq CPTT)\n",
    "## Only use clusters with modisco results\n",
    "clusters_to_use = np.unique([i.split(\"__\")[0] for i in list(pat_corrs.keys())])\n",
    "clusters_to_use = [map_r7_to_lvl4[map_r7_to_replace_dash[i]] for i in \\\n",
    "    clusters_to_use]\n",
    "\n",
    "shared_tfs = np.intersect1d(\\\n",
    "    map_id_to_tf[\"TF_Name\"].unique(), \n",
    "    cite_rna.index.values)\n",
    "\n",
    "tf2gene = pearson_corr_df_to_df(\\\n",
    "    cite_rna.loc[shared_tfs, clusters_to_use].T,\n",
    "    cite_rna.loc[gene_to_seqlets[\"Gene\"].unique(), clusters_to_use].T).dropna()\n",
    "\n",
    "tf2gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BMCP...\n",
      "\tBMCP has 255 TFs expressed...\n",
      "\t62807750 base connections.\n",
      "\t29264261 active.\n",
      "MP...\n",
      "\tMP has 247 TFs expressed...\n",
      "\t60587046 base connections.\n",
      "\t24724140 active.\n",
      "CLP1-b...\n",
      "\tCLP1-b has 273 TFs expressed...\n",
      "\t68696044 base connections.\n",
      "\t34460676 active.\n",
      "MPP1-MKP...\n",
      "\tMPP1-MKP has 286 TFs expressed...\n",
      "\t68971323 base connections.\n",
      "\t40591885 active.\n",
      "MPP1-G1...\n",
      "\tMPP1-G1 has 298 TFs expressed...\n",
      "\t68320530 base connections.\n",
      "\t40920269 active.\n",
      "ErP-1...\n",
      "\tErP-1 has 215 TFs expressed...\n",
      "\t45856367 base connections.\n",
      "\t13753497 active.\n",
      "ErP-2...\n",
      "\tErP-2 has 212 TFs expressed...\n",
      "\t43484408 base connections.\n",
      "\t13084367 active.\n",
      "ErP-HSC...\n",
      "\tErP-HSC has 250 TFs expressed...\n",
      "\t59449782 base connections.\n",
      "\t23252803 active.\n",
      "MPP2-G2...\n",
      "\tMPP2-G2 has 304 TFs expressed...\n",
      "\t71977057 base connections.\n",
      "\t47039174 active.\n",
      "MPP2-M...\n",
      "\tMPP2-M has 293 TFs expressed...\n",
      "\t68982438 base connections.\n",
      "\t42764571 active.\n",
      "MPP2-S...\n",
      "\tMPP2-S has 296 TFs expressed...\n",
      "\t71476009 base connections.\n",
      "\t42500052 active.\n",
      "MkP-HSC...\n",
      "\tMkP-HSC has 264 TFs expressed...\n",
      "\t61462045 base connections.\n",
      "\t33873308 active.\n",
      "IG2-MP...\n",
      "\tIG2-MP has 233 TFs expressed...\n",
      "\t49960802 base connections.\n",
      "\t17554929 active.\n",
      "IG2-proNeu1...\n",
      "\tIG2-proNeu1 has 244 TFs expressed...\n",
      "\t51096404 base connections.\n",
      "\t20383637 active.\n",
      "qHSC...\n",
      "\tqHSC has 326 TFs expressed...\n",
      "\t73331403 base connections.\n",
      "\t51749047 active.\n",
      "MDP-1...\n",
      "\tMDP-1 has 253 TFs expressed...\n",
      "\t55107897 base connections.\n",
      "\t22991985 active.\n",
      "MDP-2...\n",
      "\tMDP-2 has 245 TFs expressed...\n",
      "\t61324879 base connections.\n",
      "\t25049419 active.\n",
      "MEP...\n",
      "\tMEP has 266 TFs expressed...\n",
      "\t67460486 base connections.\n",
      "\t33473433 active.\n",
      "MkP-Kit...\n",
      "\tMkP-Kit has 265 TFs expressed...\n",
      "\t61903934 base connections.\n",
      "\t33772085 active.\n",
      "ML-CC...\n",
      "\tML-CC has 263 TFs expressed...\n",
      "\t66056243 base connections.\n",
      "\t31472510 active.\n",
      "HSC-Ly...\n",
      "\tHSC-Ly has 322 TFs expressed...\n",
      "\t73643547 base connections.\n",
      "\t51864508 active.\n",
      "MPP4-TCA...\n",
      "\tMPP4-TCA has 318 TFs expressed...\n",
      "\t73646137 base connections.\n",
      "\t48696586 active.\n",
      "MPP5-IER...\n",
      "\tMPP5-IER has 321 TFs expressed...\n",
      "\t73215258 base connections.\n",
      "\t48329848 active.\n",
      "MPP5 Ly-I...\n",
      "\tMPP5 Ly-I has 308 TFs expressed...\n",
      "\t71157207 base connections.\n",
      "\t46373289 active.\n",
      "ML-1b...\n",
      "\tML-1b has 265 TFs expressed...\n",
      "\t62703551 base connections.\n",
      "\t30968162 active.\n",
      "ML-3...\n",
      "\tML-3 has 256 TFs expressed...\n",
      "\t63505846 base connections.\n",
      "\t30323028 active.\n",
      "ML-MDP...\n",
      "\tML-MDP has 254 TFs expressed...\n",
      "\t61654226 base connections.\n",
      "\t27848033 active.\n",
      "ML-2...\n",
      "\tML-2 has 254 TFs expressed...\n",
      "\t59606803 base connections.\n",
      "\t26770999 active.\n",
      "MPP3-IER...\n",
      "\tMPP3-IER has 286 TFs expressed...\n",
      "\t70350105 base connections.\n",
      "\t40239758 active.\n",
      "MPP3-S...\n",
      "\tMPP3-S has 285 TFs expressed...\n",
      "\t68951500 base connections.\n",
      "\t37614789 active.\n",
      "proNeu-1a...\n",
      "\tproNeu-1a has 243 TFs expressed...\n",
      "\t52145756 base connections.\n",
      "\t21550940 active.\n",
      "aHSC...\n",
      "\taHSC has 323 TFs expressed...\n",
      "\t73195574 base connections.\n",
      "\t49684002 active.\n"
     ]
    }
   ],
   "source": [
    "dpscore_threshold = 0.05\n",
    "gex_threshold = 0.1\n",
    "gexscores_max = cite_rna.max(axis=1)\n",
    "dp_scores_max = dp_scores.max(axis=1)\n",
    "path_to_tf_links = \"output/GRN/all_genes/tf_links_observed/\"\n",
    "list_tf_link_files = os.listdir(path_to_tf_links)\n",
    "list_tfs = [i.split(\"_\")[0] for i in list_tf_link_files]\n",
    "cols_to_save = [\"Gene\", \"seqlet_idx\", \"TF_con\", \"con_score\"]\n",
    "for tmp_cluster in dp_scores.columns.values:\n",
    "    print(f\"{tmp_cluster}...\")\n",
    "    tmp_dp_scores = dp_scores[tmp_cluster].copy()\n",
    "    tmp_dp_scores.loc[tmp_dp_scores < dpscore_threshold] = 0\n",
    "    tmp_gexscores = cite_rna[tmp_cluster].copy()\n",
    "    tmp_gexscores.loc[tmp_gexscores < 0.1] = 0\n",
    "    tmp_check_tfs = tmp_gexscores.loc[list_tfs]\n",
    "    tmp_check_tfs = tmp_check_tfs.loc[tmp_check_tfs > gex_threshold]\n",
    "    print(f\"\\t{tmp_cluster} has {len(tmp_check_tfs)} TFs expressed...\")\n",
    "    # Load up the expressed TF connections\n",
    "    tmp_conns = []\n",
    "    for tmp_tf in tmp_check_tfs.index.values:\n",
    "        tmp_conns.append(pd.read_feather(os.path.join(\\\n",
    "            path_to_tf_links, f\"{tmp_tf}_links.fea\")))\n",
    "\n",
    "    tmp_conns = pd.concat(tmp_conns)\n",
    "    n_conns_pre_filter = tmp_conns.shape[0]\n",
    "\n",
    "    tmp_conns = tmp_conns.loc[~tmp_conns[\"Gene\"].str.startswith(\"Rps\")]\n",
    "    tmp_conns = tmp_conns.loc[~tmp_conns[\"Gene\"].str.startswith(\"Rpl\")]\n",
    "    tmp_conns[\"con_score\"] = (\\\n",
    "        tmp_dp_scores.loc[tmp_conns[\"seqlet_idx\"].values] * \\\n",
    "        tmp_gexscores.loc[tmp_conns[\"Gene\"].values].values * \\\n",
    "        tmp_gexscores.loc[tmp_conns[\"TF_con\"].values].values).values\n",
    "    tmp_conns = tmp_conns.loc[tmp_conns[\"con_score\"] > 0]\n",
    "    tmp_conns = tmp_conns[cols_to_save]\n",
    "    tmp_conns.reset_index(drop=True).to_feather(f\"output/GRN/all_genes/\"\\\n",
    "        \"cluster_specific_tf_links/{tmp_cluster}_conns.fea\")\n",
    "    print(f\"\\t{n_conns_pre_filter} base connections.\")\n",
    "    print(f\"\\t{tmp_conns.shape[0]} active.\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyInfinityFlow",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
